### ISSUE TITLE
ENH: Methods `previous` / `next` for `RegularGridInterpolator`

### ISSUE THREAD

**Issue by orangelynx** [scipy/scipy#18994]
### Is your feature request related to a problem? Please describe.

The function `interp1d` is marked as legacy. Although the documentation is not specific as to why and what should be used as replacement. After some digging, the reason appears to be that the function is based on FITPACK, which is considered legacy (mostly resulting from controversy around the `interp2d` function). The recommended replacement for `interp2d` (and by extension `interp1d`) is `RegularGridInterpolator`. In comparison to `interp1d` it seems to lack functionality however, especially methods `previous` and `next`. The former is e.g. frequently used for zero-order hold computations.

### Describe the solution you'd like.

While there's open questions like how `previous` and `next` are to be interpreted in higher dimensions, it seems that functionality is lost should `interp1d` be eventually removed without a suitable replacement for the `previous` and `next` interpolation modes. I suggest to think about adding these modes to RegularGridInterpolator - possibly restricted to the 1d case.

### Describe alternatives you've considered.

Rewrite `interp1d` without dependency on legacy FITPACK code.

**Comment by ev-br** [comment]
TL;DR : there is planned functionality loss, there's no need to replace anything.

Let's clear some background first. `interp1d` is declared _legacy_ not _deprecated_. This is to say   
- there is no plan to remove `interp1d`;
- we are not very interested in adding new features to `interp1d` (but this is not prohibited either, we can consider a well-motivated proposal, just the bar is quite high);
- we believe that for almost everything which `interp1d` does, there's a better way;
- there is no plan to remove `interp1d`.

`nearest`, `previous` and `next` modes are one exception to the last point. They do not have a dedicated replacement in the `scipy.interpolate` namespace. That said, they are not rocket science either: https://github.com/scipy/scipy/blob/v1.11.1/scipy/interpolate/_interpolate.py#L725

> The recommended replacement for interp2d (and by extension interp1d) is ...

There is no extension here. `interp2d` is deprecated and will be removed, and there are at least two ways to mitigate the removal: use RGI or bug-for-bug compatible FITPACK replacements, as documented at 
https://docs.scipy.org/doc/scipy-1.11.1/notebooks/interp_transition_guide.html

>  the reason <for marking interp1d legacy> appears to be that the function is based on FITPACK,

Not really. `interp1d` is not fitpack based. 

Now, for the `RGI`:  we definitely will not add anything d=1 only to RGI; just as you say, generalization to d > 1 is not obvious and so far there is no clear need to even think about it IMO.

So, the gist of this issue seems to be based on misunderstanding. 
The solution is probably to improve the documentation, and we would greatly appreciate suggestions and especially pull requests with doc improvements.
(and you are probably best positioned for such improvements @orangelynx).

**Comment by orangelynx** [comment]
> So, the gist of this issue seems to be based on misunderstanding. The solution is probably to improve the documentation
I agree, but there's at least one implementation point worth discussing (see below).

> interp1d is declared legacy not deprecated. This is to say there is no plan to remove interp1d;
In terms of documentation, this is not conveyed by the current text of the legacy note:
>This class is considered legacy and will no longer receive updates. This could also mean it will be removed in future SciPy versions.
The mere warning that it *may* be removed is enough for many developers to not use it in new code in the first place.

> nearest, previous and next modes are one exception to the last point. They do not have a dedicated replacement in the scipy.interpolate namespace. That said, they are not rocket science either.
This can be said for many functions in numpy / scipy, but code becomes cleaner if even simple computations are nicely tucked away in the responsible library. The question then is whether the use case is frequent enough to warrant implementing & maintaining it? Given the use for zero-order hold upsampling I'd welcome a discussion on this. Two preliminary thoughts on this:
- I see why removing the legacy status from `interp1d` is not desirable, as the function still has functionality that is best taken from other implementations.
- A suitable replacement is perhaps best suited to the `scipy.signal` namespace due to the nature of the use case and the 1d restriction.

I'll think about suitable ways to rephrase the doc on `interp1d`.

**Comment by ev-br** [comment]
Yes please, doc improvements would be most welcome, happy to review a PR.

**Comment by ev-br** [comment]
No activity for a month, so I'm going to close it. Still happy to review a doc improvement PR if you come back to sending one @orangelynx !
